File: 2022/3/3.py
with open('input.txt', 'r') as f:
    rucksack_items = []
    for line in f:
        rucksack_items.append(line.strip())

def get_priority(characters):
        ascii = [ord(c) for c in characters]
        priorities = []
        for a in ascii:
            if a > 96:
                priorities.append(a - 96)
            elif a < 91:
                priorities.append(a - 38)
        return sum(priorities)


def rucksack_priorities(rucksack_items):
    total = 0
    for item in rucksack_items:
        compartment_1, compartment_2 = \
            set(item[: len(item) // 2]), set(item[len(item) // 2 :])
        duplicates = compartment_1.intersection(compartment_2)
        total += get_priority(duplicates)
    return total

print(rucksack_priorities(rucksack_items))

def badges(rucksack_items, group_size=3):
    total =  0
    group = []
    for elf in rucksack_items:
        if len(group) < group_size:
            group.append(set(elf))
        if len(group) == group_size:
            badge = group[0].intersection(*group)
            total += get_priority(badge)
            group.clear()
    return total

print(badges(rucksack_items))

File: 2022/3/3.js
const fs = require('fs');

const inputFilePath = 'input.txt';
const rucksack_items = fs.readFileSync(inputFilePath, 'utf-8').trim().split('\n');

function get_priority(characters) {
    let ascii = [];
    for (let i = 0; i < characters.length; i++) {
        ascii.push(characters.charCodeAt(i));
    }
    let priorities = [];
    for (let i = 0; i < ascii.length; i++) {
        let a = ascii[i];
        if (a > 96) {
            priorities.push(a - 96);
        } else if (a < 91) {
            priorities.push(a - 38);
        }
    }
    return priorities.reduce((acc, val) => acc + val, 0);
}

function rucksack_priorities(rucksack_items) {
    let total = 0;
    for (let i = 0; i < rucksack_items.length; i++) {
        let item = rucksack_items[i];
        let mid = Math.floor(item.length / 2);
        let compartment_1 = new Set(item.substring(0, mid));
        let compartment_2 = new Set(item.substring(mid));
        let duplicates = new Set([...compartment_1].filter(x => compartment_2.has(x)));
        total += get_priority(duplicates);
    }
    return total;
}

console.log(rucksack_priorities(rucksack_items));

function badges(rucksack_items, group_size = 3) {
    let total = 0;
    let group = [];
    for (let i = 0; i < rucksack_items.length; i++) {
        let elf = rucksack_items[i];
        if (group.length < group_size) {
            group.push(new Set(elf));
        }
        if (group.length === group_size) {
            let badge = new Set([...group[0]].filter(x => group[1].has(x) && group[2].has(x)));
            total += get_priority(badge);
            group = [];
        }
    }
    return total;
}

console.log(badges(rucksack_items));
